File: backend/src/api/v1/backtest.py
"""回测结果API：查询回测历史、获取最新回测指标与净值曲线"""
from __future__ import annotations

import json
import logging
from pathlib import Path
from typing import Any

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

from src.backtest import run_backtest_and_export

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/backtest", tags=["backtest"])

PROJECT_ROOT = Path(__file__).resolve().parent.parent.parent.parent
DATA_DIR = PROJECT_ROOT / "data"
BACKTEST_RESULTS_DIR = DATA_DIR / "backtest_results"


class BacktestRunRequest(BaseModel):
    """在线触发回测请求体。"""

    model_type: str = "lightgbm"
    top_n: int = 10
    initial_capital: float = 1_000_000.0
    use_split: str = "test"
    train_end: str = "2020-12-31"
    val_end: str = "2022-12-31"
    skip_initial_weeks: int = 0


@router.get("/results")
def get_backtest_results(
    model_type: str | None = Query(default=None, description="模型类型筛选: lightgbm 或 xgboost"),
    limit: int = Query(default=10, ge=1, le=100, description="返回条数"),
) -> dict[str, Any]:
    """获取回测结果列表（从SQLite读取）"""
    try:
        from src.db import init_db, session_scope
        from src.db.models import BacktestResult

        init_db()
        results = []
        with session_scope() as session:
            query = session.query(BacktestResult).order_by(BacktestResult.id.desc())
            if model_type:
                query = query.filter(BacktestResult.model_type == model_type)
            rows = query.limit(limit).all()
            for row in rows:
                metrics = json.loads(row.metrics_json) if row.metrics_json else {}
                results.append({
                    "id": row.id,
                    "model_type": row.model_type,
                    "created_at": str(row.created_at) if row.created_at else None,
                    "metrics": metrics,
                })

        return {"code": 200, "data": {"results": results, "count": len(results)}}
    except Exception as e:
        logger.exception("查询回测结果失败")
        raise HTTPException(status_code=500, detail=str(e)) from e


@router.get("/results/{result_id}")
def get_backtest_detail(result_id: int) -> dict[str, Any]:
    """获取单条回测结果详情（含净值曲线和持仓）"""
    try:
        from src.db import init_db, session_scope
        from src.db.models import BacktestResult

        init_db()
        with session_scope() as session:
            row = session.query(BacktestResult).filter(BacktestResult.id == result_id).first()
            if not row:
                raise HTTPException(status_code=404, detail=f"回测结果 {result_id} 不存在")

            nav_data = json.loads(row.nav_json) if row.nav_json else {}
            metrics = json.loads(row.metrics_json) if row.metrics_json else {}
            params = json.loads(row.params_json) if row.params_json else {}

            return {
                "code": 200,
                "data": {
                    "id": row.id,
                    "model_type": row.model_type,
                    "created_at": str(row.created_at) if row.created_at else None,
                    "metrics": metrics,
                    "params": params,
                    "nav": nav_data,
                },
            }
    except HTTPException:
        raise
    except Exception as e:
        logger.exception("查询回测详情失败")
        raise HTTPException(status_code=500, detail=str(e)) from e


@router.get("/latest")
def get_latest_backtest() -> dict[str, Any]:
    """获取最新回测结果（含净值曲线和指标）"""
    try:
        from src.db import init_db, session_scope
        from src.db.models import BacktestResult

        init_db()
        results = {}
        with session_scope() as session:
            for model_type in ("lightgbm", "xgboost"):
                row = (
                    session.query(BacktestResult)
                    .filter(BacktestResult.model_type == model_type)
                    .order_by(BacktestResult.id.desc())
                    .first()
                )
                if row:
                    nav_data = json.loads(row.nav_json) if row.nav_json else {}
                    metrics = json.loads(row.metrics_json) if row.metrics_json else {}
                    results[model_type] = {
                        "id": row.id,
                        "metrics": metrics,
                        "nav_points": nav_data.get("nav_points", []),
                        "created_at": str(row.created_at) if row.created_at else None,
                    }

        return {"code": 200, "data": results}
    except Exception as e:
        logger.exception("查询最新回测结果失败")
        raise HTTPException(status_code=500, detail=str(e)) from e


@router.get("/comparison")
def get_backtest_comparison() -> dict[str, Any]:
    """获取双模型对比数据"""
    try:
        comparison_path = BACKTEST_RESULTS_DIR / "comparison.json"
        nav_path = BACKTEST_RESULTS_DIR / "comparison_nav.json"

        comparison = {}
        if comparison_path.exists():
            with open(comparison_path, "r", encoding="utf-8") as f:
                comparison = json.load(f)

        nav_comparison = {}
        if nav_path.exists():
            with open(nav_path, "r", encoding="utf-8") as f:
                nav_comparison = json.load(f)

        return {
            "code": 200,
            "data": {
                "comparison": comparison,
                "nav_comparison": nav_comparison,
            },
        }
    except Exception as e:
        logger.exception("查询对比数据失败")
        raise HTTPException(status_code=500, detail=str(e)) from e


@router.get("/model-info")
def get_model_analysis() -> dict[str, Any]:
    """获取模型分析数据（特征重要性、NDCG曲线等）"""
    try:
        models_dir = PROJECT_ROOT / "models"
        result: dict[str, Any] = {}

        from src.data_loader import load_factor_columns
        factor_cols = load_factor_columns(data_dir=DATA_DIR)
        fkey_map = {f"f{i}": name for i, name in enumerate(factor_cols)}

        for model_type in ("lightgbm", "xgboost"):
            fi_path = models_dir / f"{model_type}_feature_importance.json"
            metrics_path = models_dir / f"{model_type}_metrics.json"

            model_data: dict[str, Any] = {}
            if fi_path.exists():
                with open(fi_path, "r", encoding="utf-8") as f:
                    fi = json.load(f)
                mapped = {}
                for k, v in fi.items():
                    mapped[fkey_map.get(k, k)] = v
                ranked = sorted(mapped.items(), key=lambda x: -x[1])
                model_data["feature_importance"] = [
                    {"feature": k, "importance": v} for k, v in ranked
                ]
            if metrics_path.exists():
                with open(metrics_path, "r", encoding="utf-8") as f:
                    model_data["metrics"] = json.load(f)

            result[model_type] = model_data

        eval_path = models_dir / "evaluation_metrics.json"
        if eval_path.exists():
            with open(eval_path, "r", encoding="utf-8") as f:
                result["evaluation"] = json.load(f)

        return {"code": 200, "data": result}
    except Exception as e:
        logger.exception("查询模型分析数据失败")
        raise HTTPException(status_code=500, detail=str(e)) from e


@router.post("/run", summary="在线触发回测")
async def run_backtest_api(req: BacktestRunRequest) -> dict[str, Any]:
    """
    在线触发回测：执行回测、落库、导出JSON，返回结果ID和核心指标。

    参数:
        req: 回测参数请求体

    返回:
        包含 result_id、model_type、核心指标的字典

    异常:
        回测执行失败时返回HTTP 500
    """
    try:
        result = run_backtest_and_export(
            top_n=req.top_n,
            initial_capital=req.initial_capital,
            use_split=req.use_split,
            train_end=req.train_end,
            val_end=req.val_end,
            skip_initial_weeks=req.skip_initial_weeks,
        )
        model_pack = result.get(req.model_type, {})
        metrics = model_pack.get("metrics", {})
        result_df = model_pack.get("result_df")
        result_id: int | None = None
        if result_df is not None and not result_df.empty:
            from src.db import init_db, session_scope
            from src.db.models import BacktestResult

            init_db()
            with session_scope() as session:
                latest = (
                    session.query(BacktestResult)
                    .filter(BacktestResult.model_type == req.model_type)
                    .order_by(BacktestResult.id.desc())
                    .first()
                )
                if latest:
                    result_id = latest.id
        return {
            "status": "ok",
            "result_id": result_id,
            "model_type": req.model_type,
            "top_n": req.top_n,
            "metrics": metrics,
        }
    except Exception as exc:
        logger.exception("回测执行失败")
        raise HTTPException(status_code=500, detail=f"回测执行失败: {exc}") from exc

File: project/frontend/src/services/backtest.ts
/**
 * 回测与模型分析API服务
 * 对接后端回测结果查询、模型分析数据接口
 */
import { api } from './api'
import type { FeatureImportance } from '../types'

/** 回测指标 */
export interface BacktestMetricsData {
  annualized_return?: number
  sharpe_ratio?: number
  max_drawdown?: number
  cumulative_return?: number
  win_rate?: number
  turnover_rate?: number
  drawdown_peak_date?: string
  drawdown_trough_date?: string
}

/** 回测结果项 */
export interface BacktestResultItem {
  id: number
  model_type: string
  created_at: string | null
  metrics: BacktestMetricsData
}

/** 净值数据点 */
export interface NavPoint {
  date: string
  nav: number
  benchmark_nav?: number
}

/** 最新回测数据 */
export interface LatestBacktestData {
  id: number
  metrics: BacktestMetricsData
  nav_points: NavPoint[]
  created_at: string | null
}

/** 模型分析数据 */
export interface ModelAnalysisData {
  feature_importance: Array<{ feature: string; importance: number }>
  metrics: Record<string, unknown>
}

/** 模型信息响应 */
export interface ModelInfoFullResponse {
  lightgbm: ModelAnalysisData
  xgboost: ModelAnalysisData
  evaluation?: Record<string, unknown>
}

/**
 * 获取回测结果列表
 * @param modelType 模型类型筛选
 * @param limit 返回条数
 */
export async function getBacktestResults(
  modelType?: 'lightgbm' | 'xgboost',
  limit: number = 10
): Promise<{ results: BacktestResultItem[]; count: number }> {
  const params: Record<string, string | number> = { limit }
  if (modelType) params.model_type = modelType
  return api.get('/backtest/results', params)
}

/**
 * 获取最新回测数据（含净值曲线）
 */
export async function getLatestBacktest(): Promise<Record<string, LatestBacktestData>> {
  return api.get('/backtest/latest')
}

/**
 * 获取双模型对比数据
 */
export async function getBacktestComparison(): Promise<{
  comparison: Record<string, unknown>
  nav_comparison: Record<string, unknown>
}> {
  return api.get('/backtest/comparison')
}

/**
 * 获取模型分析数据（特征重要性、NDCG等）
 */
export async function getModelAnalysis(): Promise<ModelInfoFullResponse> {
  return api.get('/backtest/model-info')
}

/**
 * 获取特征重要性列表
 * @param modelType 模型类型
 */
export async function getFeatureImportance(
  modelType: 'lightgbm' | 'xgboost' = 'xgboost'
): Promise<FeatureImportance[]> {
  const data = await getModelAnalysis()
  const modelData = data[modelType]
  if (!modelData?.feature_importance) return []
  return modelData.feature_importance
    .map((fi) => ({
      feature: fi.feature,
      importance: fi.importance,
    }))
    .sort((a, b) => b.importance - a.importance)
}

/** 回测与模型服务对象 */
export const backtestService = {
  getBacktestResults,
  getLatestBacktest,
  getBacktestComparison,
  getModelAnalysis,
  getFeatureImportance,
}

export default backtestService
